### PR TITLE
introduced breaking change Failure<T, E>

### PR DESCRIPTION
**Pull request by elegon32**
this is my proposal for the feature. #12 

I did not add unit tests yet...
But this seems to work in my project...

I changed some names which did not make sense to me anymore... 
What do you think?

### CODE REVIEW

File: projects/lib/src/lib/remote-data.ts [hunk 1]
```diff
@@ -26,13 +26,24 @@ export class InProgress<T> {
...
 }
 
-export class Failure<E> {
+export class Failure<E, T> {
   private tag = RemoteDataTags.Failure;
-  private constructor(private err: E) {}
-  static of<T, E = DefaultError>(err: E): RemoteData<T, E> {
-    return new Failure(err);
+  private constructor(private err: E, private val?: T) {}
+  static of<T, E = DefaultError>(err: E, val?: T): RemoteData<T, E> {
```
Comment (joanllenas) [inline projects/lib/src/lib/remote-data.ts:32]: I like the idea of `val` being optional, but we should change that also in the `InProgress` constructors

File: projects/lib/src/lib/remote-data.ts [hunk 2]
```diff
@@ -26,13 +26,24 @@ export class InProgress<T> {
...
+    if (val) {
```
Comment (joanllenas) [inline projects/lib/src/lib/remote-data.ts]: This if/else statement is redundant, `new Failure(err, val);` does exactly the same

File: projects/lib/src/lib/remote-data.ts [hunk 3]
```diff
@@ -26,13 +26,24 @@ export class InProgress<T> {
...
+      return new Failure(err, val);
+    } else {
+      return new Failure(err);
+    }
   }
-  value(): E {
+  value(): T | undefined {
+    if (this.val) {
```
Comment (joanllenas) [inline projects/lib/src/lib/remote-data.ts]: This if/else statement is not necessary,  `return this.val;` does the same

**Review by joanllenas — CHANGES_REQUESTED** [review]
I added a couple of quick comments of things I noticed, tomorrow I'll actually download the PR and test it properly.
Thanks!

File: projects/lib/src/lib/pipes.ts [hunk 4]
```diff
@@ -129,18 +129,28 @@ export class GetInProgressPipe implements PipeTransform {
   }
 }
 
-@Pipe({ name: 'successOrInProgressValue' })
-export class GetSuccessOrInProgressValuePipe implements PipeTransform {
+@Pipe({ name: 'remoteDataValue' })
+export class GetRemoteDataPipe implements PipeTransform {
```
Comment (joanllenas) [inline projects/lib/src/lib/pipes.ts]: Let's rename this to `GetRemoteDataValuePipe` for consistency

**Review by joanllenas — CHANGES_REQUESTED** [review]
One last thing.
In the `app.module.ts` file, line 48, change:
```
            {
              matcher: (tag: string) => tag === RemoteDataTags.Failure,
              mapper: () => Failure.of(json.err)
            },
```
for
```
            {
              matcher: (tag: string) => tag === RemoteDataTags.Failure,
              mapper: () => Failure.of(json.err, json.val)
            },
```

**Comment by elegon32** [comment]
Done. I added some tests, too...
Do you change the version number accordingly to make it available via npm?
Thank you!

**Review by joanllenas — APPROVED** [review]
Looking good 👍

**Pull request merged**